### PR TITLE
MM-49030: Add metrics

### PR DESCRIPTION
**Pull request by agnivade**
- We expose prometheus metrics from Perseus
- Added golangci
- Added health endpoint

https://mattermost.atlassian.net/browse/MM-49032
https://mattermost.atlassian.net/browse/MM-49030

### CODE REVIEW

**Review by isacikgoz — APPROVED** [review]
LGTM

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+          go-version: ${{ matrix.go-version }}
+      - name: Checkout Code
+        uses: actions/checkout@v2
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v2
+        with:
+          version: v1.50
+
+          # Optional: if set to true then the action will use pre-installed Go.
+          skip-go-installation: true
```
Comment (agarciamontoro) [inline .github/workflows/golangci-lint.yml:25]: Nit: no newline at the end of file.

File: .gitignore [hunk 2]
```diff
@@ -16,4 +16,5 @@ Chart.lock
 config/config.json
 *.log
 
-
+# Ignore the binary
+perseus
```
Comment (agarciamontoro) [inline .gitignore:20]: Nit: no newline at the end of file.

File: .golangci.yml [hunk 3]
```diff
@@ -8,15 +8,23 @@ run:
...
+  govet:
+    check-shadowing: true
+    enable-all: true
+    disable:
+      - fieldalignment
+
 linters:
-  # enable some extra linters, see here for the list: https://golangci-lint.run/usage/linters/
+  disable-all: true
   enable:
```
Comment (agarciamontoro) [inline .golangci.yml:22]: Ah, that's better, thanks!

File: internal/server/health.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+func (s *Server) healthHandler(w http.ResponseWriter, r *http.Request) {
+	err := json.NewEncoder(w).Encode(healthResponse{
+		CommitHash:   CommitHash,
+		BuildVersion: BuildVersion,
+		BuildDate:    BuildDate,
+		GoVersion:    GoVersion,
+	})
+	if err != nil {
+		s.logger.Printf("failed to write health response: %v\n", err)
```
Comment (agarciamontoro) [inline internal/server/health.go:23]: Nit:
```suggestion
		s.logger.Printf("failed to write health response: %w\n", err)
```

File: internal/server/metrics.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/collectors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	namespace = "perseus"
```
Comment (agarciamontoro) [inline internal/server/metrics.go:13]: Nit: I understand this is metrics-only, but as it's in the `server` package, should we name it something more explicit like `metricsNamespace`?

File: internal/server/pool_collector.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+			nil, prometheus.Labels{"dest_host": destHost, "dest_db": destDB},
+		),
+		waitCount: prometheus.NewDesc(
+			fqName("wait_count_total"),
+			"The total number of connections waited for.",
+			nil, prometheus.Labels{"dest_host": destHost, "dest_db": destDB},
+		),
+		waitDuration: prometheus.NewDesc(
+			fqName("wait_duration_seconds_total"),
+			"The total time blocked waiting for a new connection.",
```
Comment (agarciamontoro) [inline internal/server/pool_collector.go:54]: I'm of the opinion that it's never too much information in these descriptions, even if the name is already clear:
```suggestion
			"The total time (in seconds) blocked waiting for a new connection.",
```

File: internal/server/pool_collector.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	p *Pool
+
+	maxOpenConnections *prometheus.Desc
+
+	openConnections  *prometheus.Desc
+	inUseConnections *prometheus.Desc
+	idleConnections  *prometheus.Desc
+
+	waitCount         *prometheus.Desc
+	waitDuration      *prometheus.Desc
```
Comment (agarciamontoro) [inline internal/server/pool_collector.go:15]: Can we add the unit to the name itself?
```suggestion
	waitDurationSec      *prometheus.Desc
```

File: internal/server/pool_collector.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+func (c *poolCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.maxOpenConnections
+	ch <- c.openConnections
+	ch <- c.inUseConnections
+	ch <- c.idleConnections
+	ch <- c.waitCount
+	ch <- c.waitDuration
+	ch <- c.maxIdleClosed
+	ch <- c.maxLifetimeClosed
+	ch <- c.maxIdleTimeClosed
```
Comment (agarciamontoro) [inline internal/server/pool_collector.go:85]: Nit level over 9000: we should maintain the same order
```suggestion
	ch <- c.maxIdleClosed
	ch <- c.maxIdleTimeClosed
	ch <- c.maxLifetimeClosed
```

File: internal/server/pool_collector.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	stats := c.p.Stats()
+	ch <- prometheus.MustNewConstMetric(c.maxOpenConnections, prometheus.GaugeValue, float64(stats.MaxOpenConnections))
+	ch <- prometheus.MustNewConstMetric(c.openConnections, prometheus.GaugeValue, float64(stats.OpenConnections))
+	ch <- prometheus.MustNewConstMetric(c.inUseConnections, prometheus.GaugeValue, float64(stats.InUse))
+	ch <- prometheus.MustNewConstMetric(c.idleConnections, prometheus.GaugeValue, float64(stats.Idle))
+	ch <- prometheus.MustNewConstMetric(c.waitCount, prometheus.CounterValue, float64(stats.WaitCount))
+	ch <- prometheus.MustNewConstMetric(c.waitDuration, prometheus.CounterValue, stats.WaitDuration.Seconds())
+	ch <- prometheus.MustNewConstMetric(c.maxIdleClosed, prometheus.CounterValue, float64(stats.MaxIdleClosed))
+	ch <- prometheus.MustNewConstMetric(c.maxLifetimeClosed, prometheus.CounterValue, float64(stats.MaxLifetimeClosed))
+	ch <- prometheus.MustNewConstMetric(c.maxIdleTimeClosed, prometheus.CounterValue, float64(stats.MaxIdleTimeClosed))
```
Comment (agarciamontoro) [inline internal/server/pool_collector.go:99]: Same as above
```suggestion
	ch <- prometheus.MustNewConstMetric(c.maxIdleClosed, prometheus.CounterValue, float64(stats.MaxIdleClosed))
	ch <- prometheus.MustNewConstMetric(c.maxIdleTimeClosed, prometheus.CounterValue, float64(stats.MaxIdleTimeClosed))
	ch <- prometheus.MustNewConstMetric(c.maxLifetimeClosed, prometheus.CounterValue, float64(stats.MaxLifetimeClosed))
```

File: internal/server/server.go [hunk 10]
```diff
@@ -58,7 +74,30 @@ func New(cfg config.Config) (*Server, error) {
...
+		}
+		router.HandleFunc("/health", s.healthHandler)
+		router.Handle("/metrics", s.metrics.metricsHandler())
+
+		s.metricsWg.Add(1)
+		go func() {
+			defer s.metricsWg.Done()
+			err2 := s.metricsSrv.ListenAndServe()
+			if err2 != nil && !errors.Is(err2, http.ErrServerClosed) {
+				s.logger.Printf("Error while running metrics server: %v\n", err2)
```
Comment (agarciamontoro) [inline internal/server/server.go:95]: I only have nits for you today
```suggestion
				s.logger.Printf("Error while running metrics server: %w\n", err2)
```

File: internal/server/server.go [hunk 11]
```diff
@@ -137,4 +176,14 @@ func (s *Server) Stop() {
 	if err := s.poolMgr.Close(); err != nil {
 		s.logger.Printf("Error closing pool manager: %v\n", err)
 	}
+
+	if s.metricsSrv != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		if err := s.metricsSrv.Shutdown(ctx); err != nil {
+			s.logger.Printf("Error closing metrics server: %v\n", err)
```
Comment (agarciamontoro) [inline internal/server/server.go:184]:
```suggestion
			s.logger.Printf("Error closing metrics server: %w\n", err)
```

**Review by agarciamontoro — APPROVED** [review]
Cool! I left a handful of comments, but each and every one of them is a nit :nerd_face:

File: internal/server/health.go (hunk 4 above)
Comment (agnivade) [reply internal/server/health.go:23]: This is fine. We are logging it. Not sending it upwards.

File: internal/server/metrics.go (hunk 5 above)
Comment (agnivade) [reply internal/server/metrics.go:13]: This is actually used in multiple places now. I'll rename it to `serviceName`

File: internal/server/pool_collector.go (hunk 8 above)
Comment (agnivade) [reply internal/server/pool_collector.go:85]: oh dear .. you are going super saiyyan today with your nits :joy:

File: internal/server/server.go (hunk 10 above)
Comment (agnivade) [reply internal/server/server.go:95]: As above, this is fine.

File: internal/server/server.go (hunk 11 above)
Comment (agnivade) [reply internal/server/server.go:184]: Same

File: internal/server/pool_collector.go (hunk 8 above)
Comment (agarciamontoro) [reply internal/server/pool_collector.go:85]: They're nits, so feel free to ignore them!

**Comment by agnivade** [comment]
@agarciamontoro / @isacikgoz - Sorry it needs another review because I pushed another commit.

**Review by isacikgoz — APPROVED** [review]

**Pull request merged**